### PR TITLE
`Ares`: ArchUnit Filesystem Access tests for pre-compile and post-compile mode

### PR DESCRIPTION
**Pull request by sarpsahinalp**
#### General

- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Ares
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

This PR aims to introduce architectural tests to prevent access to the filesystem. It proposes an abstract method of customizing the architectural tests using a configuration file. This approach makes the Ares framework compatible with different programming languages in the future, thereby making Ares independent of the programming language used and making it easier for the instructors to customize the architectural constraints without knowing the programming language. Architectural tests helps us pinpoint the location of the code and provide precise feedback to the student.

### Description

- **New Features**
	- Introduced centralized storage for architecture test cases.
	- Added a new class for security rules related to file system interactions.
	- Implemented a mechanism for dynamically retrieving class locations based on annotations.
	- Created a custom annotation for specifying compiled student class paths.
	- Established architectural rules to prevent unauthorized file system access.

- **Bug Fixes**
	- Enhanced clarity and adherence to naming conventions in enumerations.

- **Documentation**
	- Added comprehensive architectural rules definitions in new files.

#### Code Review
- [x] Code Review 1
- [ ] Code Review 2

## Summary by CodeRabbit

- **New Features**
	- Introduced a new class for architecture testing specific to Java.
	- Added a custom annotation to manage file access permissions dynamically.
	- Launched a configuration file to set constraints for student interactions during exercises.
	- Created a centralized repository for file path constants.
	- Implemented a utility class for managing security test cases in Java.

- **Improvements**
	- Enhanced the parsing and initialization of security test cases based on defined policies.
	- Streamlined version management for dependencies.

- **Bug Fixes**
	- Improved enforcement of architectural rules to restrict unauthorized file system access by student submissions.

### CODE REVIEW

File: src/main/resources/archunit/files/java/pre-compile-layout.txt [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package %s;
```
[the remark this answers is not in the document]
Comment (sarpsahinalp) [reply src/main/resources/archunit/files/java/pre-compile-layout.txt:1]: This is a text layout

File: src/main/resources/archunit/files/java/pre-compile-layout.txt [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import de.tum.cit.ase.ares.api.architecturetest.java.postcompile.TransitivelyAccessesMethodsCondition;
+\s
+import java.util.*;
+\s
+/**
+* This class executes the security rules on the architecture.
+*/
+@StudentCompiledClassesPath("%s")
+@AnalyzeClasses(locations = PathLocationProvider.class, importOptions = ImportOption.DoNotIncludeTests.class)
+public class ArchitectureTestCase {
```
[the remark this answers is not in the document]
Comment (sarpsahinalp) [reply src/main/resources/archunit/files/java/pre-compile-layout.txt:23]: This is a layout

**Review by MarkusPaulsen — APPROVED** [review]
Code looks good.

**Review by az108 — APPROVED** [review]
LGTM

**Pull request merged**